<div style="border:solid Chocolate 2px; padding: 40px">

Привет Дмитрий меня зовут Саша Атаков. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех👍:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<b><font color='orange'>Вот тебе полезная ссылка, вдруг подчерпнешь что-то новое для себя):</font></b>

* О логистической регрессии: https://habr.com/ru/company/io/blog/265007/
* Ну и случайный лес: https://alexanderdyakonov.wordpress.com/2016/11/14/%D1%81%D0%BB%D1%83%D1%87%D0%B0%D0%B9%D0%BD%D1%8B%D0%B9-%D0%BB%D0%B5%D1%81-random-forest/
* О dummy model - https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html
* Более подробно о dummy model - https://www.geeksforgeeks.org/ml-dummy-classifiers-using-sklearn/
    
<b><font color='orange'>Общее впечатление</font></b>
* Большое спасибо за проделанную работу. Видно, что ты отлично освоил тему.
* Отлично, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* В работе я оставил несколько советов. Буду рад, если ты учтешь их.
* В конце прописал свой комментарий по всему проекту, ознакомься с ним, пожалуйста)
* С радостью отвечу на твои вопросы, касательно выданной тебе темы, если они есть. Жду новую версию проекта.)

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

### Импортируем необходимые библиотеки и загружаем датасет

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt 
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.dummy import DummyClassifier
from sklearn.neighbors import KNeighborsClassifier

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Хороший шаг - импортировать все необходимые библиотеки перед работой с проектом, ведь так ты точно не запутаешься в том, какие библиотеки ты уже импортировал, а какие стоит добавить. Молодец!
</div>

In [4]:
try:
    data = pd.read_csv('/datasets/users_behavior.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv')

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Загрузка данных по двум сценариям, удобно)
</div>

### Выведим первые 20 строк датасета 

In [3]:
data.head(20)

calls  minutes  messages   mb_used  is_ultra
0    40.0   311.90      83.0  19915.42         0
1    85.0   516.75      56.0  22696.96         0
2    77.0   467.66      86.0  21060.45         0
3   106.0   745.53      81.0   8437.39         1
4    66.0   418.74       1.0  14502.75         0
5    58.0   344.56      21.0  15823.37         0
6    57.0   431.64      20.0   3738.90         1
7    15.0   132.40       6.0  21911.60         0
8     7.0    43.39       3.0   2538.67         1
9    90.0   665.41      38.0  17358.61         0
10   82.0   560.51      20.0   9619.53         1
11   45.0   344.32      13.0  19898.81         0
12   51.0   437.13      61.0  21523.58         0
13   56.0   433.07      16.0  16702.36         0
14  108.0   587.90       0.0  14406.50         1
15    6.0    22.13       0.0   2710.09         0
16    2.0    18.73       0.0    588.89         0
17   26.0   163.62       4.0  16870.34         0
18   79.0   532.62      90.0  19908.31         0
19   49.0   341.67      81.0  11770.28         1

### Выведим общую информацию про датасет

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


###  Посмотрим на статистическую информацию по столбцам

In [5]:
data.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

### Изменим типы данных

In [6]:
data['calls'] = data['calls'].astype('int')
data['messages'] = data['messages'].astype('int')

**Вывод** <br>
Видем, что в данных нету пропусков, данных довольно много, нету аномальных значений. Так же мы изменили типы данных в некоторых столбцах на более подходящие, а значит, можно переходить к исследованию

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, абсолютно верно, данные готовы к работе. Единственное что я бы добавил еще - можно типы данных для float, которые по своему виду не похожи на float перевести в int
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Добавил изменение типов и немного поменял вывод
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Круто, молодец!
</div>

## Разбейте данные на выборки

### Разделим данные на обучающую, валидационную и тестовую выборки в соотношении 3:1:1

In [5]:
data_train, data_check = train_test_split(data, test_size=0.4, random_state=12345)
data_valid, data_test = train_test_split(data_check, test_size=0.5, random_state=12345)

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Вот это мне нравится, разделение на три выборки, молодец! Также прописал грамотный вывод по задаче, так держать!
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Слушай, можно сделать разбиение 80/10/10, обычно я так делаю, но тут видно, что данных маловато, так что для начала и 60/20/20 пойдет). А так, это не догма и в будущих проектах ты сможешь убедиться в этом)
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Спасибо за совет
</div>

### Разделим данные по признакам и целевому признаку

In [6]:
features_train = data_train.drop(['is_ultra'], axis=1)
target_train = data_train['is_ultra']

features_valid = data_valid.drop(['is_ultra'], axis=1)
target_valid = data_valid['is_ultra']

features_test = data_test.drop(['is_ultra'], axis=1)
target_test = data_test['is_ultra']

In [7]:
target_train

3027    0
434     0
1226    0
1054    0
1842    0
       ..
2817    1
546     1
382     1
2177    1
482     1
Name: is_ultra, Length: 1928, dtype: int64

**Вывод** <br>
Разбили данные на обучающую, валидационную и тестовую выборки в соотношении 3:1:1. Раделили данные по признакам и целевому признаку

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Признаки и таргеты ты подготовил, пошли дальше
</div>

## Исследуйте модели

### Модель "Решающее дерево"

Исследуем модель "Решающее дерево". С помощью цикла будем менять гиперпараметр max_depth от 1 до 10, и посмотрим в каком случае будет лучший показатель accuracy

In [9]:
best_model_tree = None
best_result_tree = 0
best_depth_tree = 0
best_criterion = None

criterions = {'gini', 'entropy'}

for crit in criterions:
    for depth in range(1, 11):
        model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion=crit) 
        model_tree.fit(features_train, target_train) 
        predictions = model_tree.predict(features_valid) 
        result = accuracy_score(target_valid, predictions) 
        if result > best_result_tree:
            best_model_tree = model_tree
            best_result_tree = result
            best_depth_tree = depth
            best_criterion = crit
        
print(f'Accuracy наилучшей модели на валидационной выборке: {best_result_tree}, при max_depth = {best_depth_tree}, \
при criterion = {best_criterion}') 

Accuracy наилучшей модели на валидационной выборке: 0.7853810264385692, при max_depth = 3, при criterion = entropy


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, дерево хорошо обучилось)
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Тут можно еще проверить гиперпараметр - атрибут разбиения (criterion), так как она осуществляет важную функцию - при обучении множества в узле, данный атрибут следит за тем, чтобы в этих множествах было меньше дисбаланса классов значений, к которым мы и предсказываем что-либо. Ну это так, на подумать, может в будущем ты захочешь его применить)
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Проверил гиперпараметр атрибут разбиения
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Я нашёл, что ещё бывает вариант этого атрибута - "log_loss", однако если я его добавлял, появлялась ошибка. Подскажите пожалуйста, почему?
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔 V2:</b> 

Смотрим, чтобы вычислить log_loss, тебе нужно знать вероятности прогнозов твоей модели (с помощью параметра predict_proba). Если ты предоставляешь только прогнозируемые классы (классы с максимальной вероятностью), эта метрика не может быть вычислена.
    
Например вот так ничего работать не будет: log_loss(lg_y_test, lg_y_predictions)
</div>

### Модель "Случайный лес"

Исследуем модель "Случайный лес". С помощью цикла будем менять гиперпараметр max_depth от 1 до 10, и гиперпараметр n_estimators от 1 до 15. Найдём лучшие значения этих гиперпараметров и выведим лучшее accuracy

In [10]:
best_model_forest = None
best_result_forest = 0
best_depth_forest = 0
best_n_estimators_forest = 0
best_criterion = None

criterions = {'gini', 'entropy'}

for crit in criterions:
    for depth in range(1, 11):
        for est in range(1, 51):
            model_forest = RandomForestClassifier(random_state=12345, max_depth = depth, n_estimators=est, criterion=crit) 
            model_forest.fit(features_train, target_train) 
            result = model_forest.score(features_valid, target_valid) 
            if result > best_result_forest:
                best_model_forest = model_forest
                best_result_forest = result
                best_depth_forest = depth
                best_n_estimators_forest = est
                best_criterion = crit

print(f'Accuracy наилучшей модели на валидационной выборке: {best_result_forest}, при max_depth = {best_depth_tree}, \
при n_estimators = {best_n_estimators_forest}, при criterion = {best_criterion}') 

Accuracy наилучшей модели на валидационной выборке: 0.8087091757387247, при max_depth = 3, при n_estimators = 40, при criterion = gini


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Лес тоже сформирован и дал лучший результат (на данный момент), отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

Деревьев стоит выбрать побольше, так как именно от этого параметра зависит качество предсказания нашей модели, ведь лес - это множество моделей дерева, где учитывается каждое дерево, дающее свой результат. Попробуй выбрать значения побольше и шаг тоже добавь
    
А так можно еще взять гиперпараметр дерева - атрибут разбиения, о котором я выше говорил
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Увеличил количество деревьеев с 15 до 50. Добавил гиперпараметр атрибут разбиения. 
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Больше увеличивать количество деревьев не стал, т.к. загрузка бы стала слишком долгой, при этом точность увеличивается лишь на несколько сотых, что, как мне кажется, в данном случае не очень важно. А вот про шаг немного не понял, можете пожалуйста объяснить?
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Смотри, шаг это то, через сколько значений твоя модель или функция итп будет переходить к следующему значению: for est in range(1, 51, 10) - третье значение будет являться шагом, который будет говорить нашей функции чтобы после 1, мы брали 1+10 значение, потом еще больше на 10 итп. Вот так вот
</div>

### Модель "Логистическая регрессия"

Исследуем модель "Логическая регрессия". Здесь нету гиперпараметров, которые бы влияли на результат.

In [11]:
best_model_lr = None
best_result_lr = 0
best_solver = None
best_c = 0

solvers = {'liblinear', 'lbfgs'}

for c in range(1, 11):
    for sol in solvers:
        model_lr = LogisticRegression(random_state=12345, solver=sol, C=c)
        model_lr.fit(features_train, target_train)
        result = model_lr.score(features_valid, target_valid)
        if result > best_result_lr:
                    best_model_lr = model_lr
                    best_result_lr = result 
                    best_solver = sol
                    best_c = c
                

print(f'Accuracy модели на валидационной выборке: {best_result_lr}, при solver = {best_solver}, при c = {best_c}')

Accuracy модели на валидационной выборке: 0.7573872472783826, при solver = liblinear, при c = 5


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Эх, логистическая регрессия показала не самый лучший результат. Возможно проблема в данных
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 

К сожалению я не увидел здесь перебор гиперпараметров, а у логистической регрессии есть очень хорошие гиперпараметры, меняя которые можно немного поднять качество: сила регуляризации и веса признаков. Попробуй поизучать данные гиперпараметры и перебрать их) 
    
Если качество не улучшится, не стоит огорчаться, у нас данных очень мало
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Добавил гиперпараметры solver и C.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Другие гиперпараметры не добовлял либо потому что не понял, как они работают, либо потому что результат не менялся или менялся в худшую сторону. При этом, удалось увеличить результат - с 0.71 до 0.757, то есть практически на 5%.  
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Ну как раз самый классный гиперпараметр - сила регуляризации. Меняя лишь данное значение, мы можем улучшить качество нашей модели на порядок
</div>

**Вывод** <br>
- Лучше всего на валидационной выборке себя показала модель "Случайный лес". У неё получился наилучший показатель accuracy - 0.8. 
- На втором месте оказалась модель "Решающее дерево" с показателем accuracy - 0.76. 
- Ну и хуже всех себя проявила модель "Логистическая регрессия" с показателем accuracy - 0.75.

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Да, все верно, случайный лес явный победитель, это круто. Пойдем же проверим ее на тестовых данных
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Немного изменил вывод с учётом изменений в предыдущих заданиях
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Понял, принял
</div>

## Проверьте модель на тестовой выборке

In [12]:
test_result = best_model_forest.score(features_test, target_test)

print(f'Accurancy модели "Случайный лес" на тестовой выборке: {test_result}')

Accurancy модели "Случайный лес" на тестовой выборке: 0.7962674961119751


**Вывод** <br>
Точность лучшей модели "Случайный лес" на тестовой выборке составила 0.796, что является довольно хорошим результатом

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
Отлично, на тестовых данных результат хороший, задача выполнена, молодец)
</div>

## (бонус) Проверьте модели на адекватность

### Первый способ

Чтобы оценить вменяемость нашей модели, сравним её результат со случайными предсказаниями. 

In [13]:
random_answers = np.random.randint(0, 2, size=643)

accuracy_random_test = accuracy_score(target_test, random_answers)
accuracy_random_test

0.49611197511664074

**Вывод** <br>
Как мы видим, случайные предсказания имеют показатель схожести с тестовой выборкой около 0.5, значит наша модель, имеющая результат около 0.8, вполне точная, а значит адекватная

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 
    
И мы видим, что случайные предсказания дают лишь 50% качества, в то время как наша лучшая модель дала 79%. Это значит, что наша модель адекватна
</div>

<div class="alert alert-block alert-info">
<b>Совет🤔:</b> 
    
На самом деле ты можешь использовать другой алгоритм для проверки адекватности на своих данных, например, Dummy Models для более удобной и функциональной проверки. Можешь почитать на досуге
    
Также хотелось бы увидеть в конце итоговый вывод о том, что ты сделал на протяжении всего проекта - краткий экскурс скажем так. Добавь его, если есть время
</div>

### Второй способ (Dummy Models)

In [29]:
strategies = ['most_frequent', 'stratified', 'uniform', 'constant']
  
test_scores = []

for s in strategies:
    if s =='constant':
        dclf = DummyClassifier(strategy = s, random_state = 12345, constant=1)
    else:
        dclf = DummyClassifier(strategy = s, random_state = 12345)
    dclf.fit(features_train, target_train)
    score = dclf.score(features_test, target_test)
    test_scores.append(score)

print(f'При strategy = {strategies[0]}, score = {test_scores[0]}')
print(f'При strategy = {strategies[1]}, score = {test_scores[1]}')
print(f'При strategy = {strategies[2]}, score = {test_scores[2]}')
print(f'При strategy = {strategies[3]}, score = {test_scores[3]}')

При strategy = most_frequent, score = 0.6842923794712286
При strategy = stratified, score = 0.536547433903577
При strategy = uniform, score = 0.4821150855365474
При strategy = constant, score = 0.3157076205287714


<div class="alert alert-block alert-warning">
<b>Изменения:</b> Вроде добавил способ Dummy Models
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Не до конца разобрался, как именно работает метод. Но, как я понял, надо сравнить эти значения с полученным значением нашей модели и если они меньше, то наше модель адекватная
</div

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Верно, ты правильно понял. Однако смотри, как работает dummy model - он берет дисбаланс значений в датасете, и берет количество вероятных ответов на 1 или 0 (в случае задач бинарной классификации). например если у нас уровень значений 1 и 0 равен 70 к 30, то среднее значение ответов 1 равно 70%, то есть уровень адекватности - 70%. Если модель переходит данный рубеж, то она адекватна, ниже - значит есть проблемы в качестве данной модели. Вот так вот! 
</div>

## Вывод

**В ходе исследования было сделано:** <br>
1. Ознакомились с данными, посмотрели на общую информацию, поменяли типы данных в некоторых столбцах
2. Разделили данные на обучающую, валидационную и тестовую выборки
3. Разделили данные на признаки и таргеты
4. Исследовали модель "Решающее дерево"
5. Исследовали модель "Случайный лес"
6. Исследовали модель "Логистическая регрессия"
7. Проверили лучшию модель "Случайный лес" на тестовой выборке
8. Проверили модель на адекватность

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Добавил вывод
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 V2:</b> 
    
Можно и так)
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<div style="border:solid Chocolate 2px; padding: 40px">
    
<h2><b>Общий вывод по проекту</b></h2>
    
Ты очень постарался, спасибо тебе за такую работу!)

В ходе проекта были выполнены все этапы проекта, главная цель достигнута. Конечно присутствует ряд моментов, которые необходимо рассмотреть, но ничего страшного, для это я здесь). Давай пройдемся по всем этапам твоего проекта чтобы понять, как ты справился:
    
<b>`Загрузка данных`</b>. С этим этапом никаких вопросов нет, здесь все сделано правильно: библиотеки загружены в отдельной ячейке над всем проектом, создан датасет, был выполнен базовый просмотр данных.
    
<b>`Исследуйте модели`</b>. На данном этапе ты реализовал три алгоритма МО:
    
- Дерево - самый базовый алгоритм, но иногда дающий приятные результаты. В нем ты попробовал гиперпараметры такие как: глубина. Довольно хороший гиперпараметр, ведь так мы проверяем большее количество факторов при определении ответа для дерева. Но можно проверить еще атрибут разбиения, попробуй это сделать)
    
- Случайный лес - самый популярный алгоритм за счет своей точности, ведь решение принимает несколько деревьев. Ты проверил гиперпараметр количество деревьев и глубина с дерева. Да, количество деревьев хорошо улучшает точность прогноза, но и повышает время обучения и принятия решения. В твоем проекте стоит попробовать большее количество деревьев, ведь от 10 до 100 это ниже дефолтного состояния леса (а он равняется 100). Попробуй другие значения. Также можешь взять дополнительные гиперпараметры, например атрибут разбиения
    
- Логистическая регрессия - самый математичный алгоритм, который хорошо показывает обучение и предсказание моделей МО со стороны математики. Ты тут проверил гиперпараметр алгоритм оптимизации, неплохо. Но чаще проверяют именно силу регуляризации. Посмотри что это такое (долгая тема, предупреждаю сразу) и реализуй ее в своем проекте, будет интересно)
    
<b>`Проверка моделей`</b>. Тут ты проверил лучшую модель на тестовых значениях, и узнал, что задача решена - результат выше требуемого. Также твоя проверка на адекватность меня довольно сильно порадовала, молодец). Однако если хочешь реализовать проверку с помощью алгоритма, то изучи материал касательно dummy models)
    
Вот так вот. Работа хорошая, у тебя все ок, но если хочешь попробовать переработать проект по моим советам, то направляю проект. Если нет возможности доработать — присылай в текущем виде и я его приму.

<div style="border:solid Chocolate 2px; padding: 40px">
    
<h2><b>Общий вывод по проекту V2</b></h2>
    
Ты прислушался ко всем моим рекомендациям - снимаю шляпу, ты молодец). проект готов к приему, но я отправляю тебе как ответы на твои вопросы. Удачи в следующих проектах!

<div style="border:solid Chocolate 2px; padding: 40px">
    
<h2><b>Общий вывод по проекту V3</b></h2>
    
Все ок, принимаю проект!